### PR TITLE
C-010: hashing password

### PR DESCRIPTION
**Pull request by letier3110**
https://www.notion.so/madfissolutions/C-010-Poor-memory-management-of-secrets-eb433138ed4c46e682f2ee6928595e10

### CODE REVIEW

**Review by LendiHop — APPROVED** [review]

**Review by andrewmarushko — APPROVED** [review]

File: src/utils/crypto.util.ts [hunk 1]
```diff
@@ -53,6 +53,8 @@ export const encryptString$ = (value: string, password: string): Observable<Encr
   );
 };
 
+export const encryptPass$ = (password: string) => pbkdf2$(password, generateSalt());
```
Comment (vixentael) [inline src/utils/crypto.util.ts:56]: that's hashing, not encryption

**Review by vixentael — CHANGES_REQUESTED** [review]
overall, good. but be careful with naming. pbkdf2 results in password _hash_, not encrypted password.

File: src/utils/crypto.util.ts (hunk 1 above)
Comment (vixentael) [reply src/utils/crypto.util.ts:56]:
```
const pbkdf2$ = (password: BinaryLike, salt: BinaryLike) =>
  from<string>(NativeModules.Aes.pbkdf2(password, salt, 5000, 256));
```

too few rounds. According to the latest NIST recommendations and OWASP guidelines, for PBKDF2 with SHA-512, 120000 rounds should be used.

**Comment by vixentael** [comment]
also, be careful, as right now PBKDF2 uses only 5000 rounds which is too low. See C-001.

File: src/utils/crypto.util.ts (hunk 1 above)
Comment (unixvb) [reply src/utils/crypto.util.ts:56]: @vixentael does `react-native-themis` provide hashing method, so we could get rid of `react-native-aes`?

File: src/shelter/shelter.ts [hunk 2]
```diff
@@ -21,14 +28,14 @@ import { getDerivationPath, getPublicKeyAndHash$, seedToPrivateKey } from '../ut
 const EMPTY_PASSWORD = '';
 
 export class Shelter {
-  private static _password$ = new BehaviorSubject(EMPTY_PASSWORD);
+  private static _hash$ = new BehaviorSubject(EMPTY_PASSWORD);
```
Comment (unixvb) [inline src/shelter/shelter.ts:31]: rename to `_passwordHash` plz

File: src/utils/crypto.util.ts (hunk 1 above)
Comment (unixvb) [reply src/utils/crypto.util.ts:56]: @letier3110 use `scrypt` instead of `pbkdf2`

Comment (vixentael) [reply src/utils/crypto.util.ts:56]: +1 to @unixvb

File: src/utils/crypto.util.ts [hunk 3]
```diff
@@ -1,16 +1,45 @@
...
+const scrypt$ = async (password: string, salt: string) =>
+  from<string>(
+    await scrypt(
+      Buffer.from(password, 'base64').toString(),
+      Buffer.from(salt, 'base64').toString(),
+      65536,
+      8,
+      1,
+      32,
+      'base64'
```
Comment (vixentael) [inline src/utils/crypto.util.ts:24]: 👀

File: src/utils/crypto.util.ts [hunk 4]
```diff
@@ -1,16 +1,45 @@
...
+    )
+  );
+
+export const hashPassword$ = (password: string) => {
+  const genSalt = async () => {
+    const salt = await symmetricKey64();
```
Comment (vixentael) [inline src/utils/crypto.util.ts:30]: 👍

**Review by vixentael — COMMENTED** [review]
👀

File: src/utils/crypto.util.ts (hunk 3 above)
Comment (letier3110) [reply src/utils/crypto.util.ts]: @vixentael  is everything ok in scrypt-call signature?

**Comment by vixentael** [comment]
@letier3110 if this PR ready for review? If so, I'll take a look.

**Comment by letier3110** [comment]
> @letier3110 if this PR ready for review? If so, I'll take a look.

yes, its ready

File: src/utils/crypto.util.ts [hunk 5]
```diff
@@ -1,16 +1,35 @@
...
-import { catchError } from 'rxjs/operators';
+import { map, switchMap, catchError } from 'rxjs/operators';
 
 export interface EncryptedData {
   encrypted64: string;
   salt64: string;
 }
 
+const scrypt$ = async (password: string, salt: string) =>
+  from<string>(await scrypt(password, salt, 65536, 8, 1, 32, 'base64'));
```
Comment (vixentael) [inline src/utils/crypto.util.ts:16]: number of rounds looks good, however, i urge you to test it on different devices, as the processing might be too slow, especially on androids. if it's too slow, try decreasing rounds to 32768 or even 16384.

see https://cryptobook.nakov.com/mac-and-key-derivation/scrypt

**Comment by vixentael** [comment]
LGTM but please test on devices. @letier3110

**Review by vixentael — APPROVED** [review]
sha256 ok

**Pull request merged**